src/pybindings/pycontroller/ChipDeviceCtrl.py

```diff
@@ -0,0 +1,320 @@
+
+    def OnStatusUpdate(self, status):
+        print("OnStatusUpdate Status: {}".format(status))
+
+    def OnPairingDeleted(self, error):
+        print("OnPairingDeleted Error: {}".format(error))
+
+
+
+def _singleton(cls):
+    instance = [None]
+
+    def wrapper(*args, **kwargs):
+        if instance[0] is None:
+            instance[0] = cls(*args, **kwargs)
+        return instance[0]
+
+    return wrapper
+
+
+class DCState(enum.IntEnum):
+    NOT_INITIALIZED = 0
+    IDLE = 1
+    BLE_READY = 2
+    RENDEZVOUS_ONGOING = 3
+    RENDEZVOUS_CONNECTED = 4
+
+@_singleton
+class ChipDeviceController(object):
+    def __init__(self, startNetworkThread=True, controllerNodeId=0, bluetoothAdapter=None):
+        self.state = DCState.NOT_INITIALIZED
+        self.devCtrl = None
+        if bluetoothAdapter is None:
+            bluetoothAdapter = 0
+        # self._ChipStack = ChipStack(bluetoothAdapter=bluetoothAdapter)
+        # self._dmLib = None
+
+
+        # devCtrl = c_void_p(None)
+        # res = self._dmLib.pychip_DeviceController_NewDeviceController(
+        #     pointer(devCtrl), controllerNodeId)
+        # if res != 0:
+        #     raise self._ChipStack.ErrorToException(res)
+
+        # self.devCtrl = devCtrl
+        # self._ChipStack.devCtrl = devCtrl
+
+        # self._Cluster = ChipClusters(self._ChipStack)
+
+        self.devCtrl = CHIPController.CHIPDeviceControllerPyBind()
+        device_address_delegate = PyDeviceAddressUpdateDelegate()
+        device_pairing_delegate = PyDevicePairingDelegate()
+        self.devCtrl.SetDeviceAddressUpdateDelegate(device_address_delegate)
+        self.devCtrl.SetDevicePairingDelegate(device_pairing_delegate)
+        self.devCtrl.Init()
+
+        def HandleKeyExchangeComplete(err):
+            if err != 0:
+                print("Failed to establish secure session to device: {}".format(err))
+                self._ChipStack.callbackRes = self._ChipStack.ErrorToException(
+                    err)
+            else:
+                print("Secure Session to Device Established")
+                self._ChipStack.callbackRes = True
+            self.state = DCState.IDLE
+            self._ChipStack.completeEvent.set()
+
+        def HandleAddressUpdateComplete(nodeid, err):
+            if err != 0:
+                print("Failed to update node address: {}".format(err))
+            else:
+                print("Node address has been updated")
+            self.state = DCState.IDLE
+            self._ChipStack.callbackRes = err
+            self._ChipStack.completeEvent.set()
+
+        self.state = DCState.IDLE
+
+    def __del__(self):
+        if self.devCtrl != None:
+            self.devCtrl.Shutdown()
+            self.devCtrl = None
+
+    def IsConnected(self):
+        return self._ChipStack.Call(
+            lambda: self._dmLib.pychip_DeviceController_IsConnected(
+                self.devCtrl)
+        )
+
+    def ConnectBle(self, bleConnection):
```

Comment:
Could you also add BLE scan functions so that we can scan the surrounding matter ble devices and get their adv infos.

Reply:
Python side of this is not complete. We need this merged in so other can contribute to the python side of things. Please see the document above.